src/SubtitleOctopus.cpp

```diff
@@ -79,18 +82,71 @@ const float MAX_UINT8_CAST = 255.9 / 255;
 
 #define CLAMP_UINT8(value) ((value > MIN_UINT8_CAST) ? ((value < MAX_UINT8_CAST) ? (int)(value * 255) : 255) : 0)
 
+typedef struct RenderBlendPart {
+    int dest_x, dest_y, dest_width, dest_height;
+    unsigned char *image;
+    RenderBlendPart *next;
+} RenderBlendPart;
+
 typedef struct {
     int changed;
     double blend_time;
-    int dest_x, dest_y, dest_width, dest_height;
-    unsigned char* image;
+    RenderBlendPart *part;
 } RenderBlendResult;
 
+// maximum regions - a grid of 3x3
+#define MAX_BLEND_STORAGES (3 * 3)
+typedef struct {
+    RenderBlendPart part;
+    ReusableBuffer buf;
+    bool taken;
+} RenderBlendStorage;
+
 typedef struct {
     double eventFinish, emptyFinish;
     int is_animated;
 } EventStopTimesResult;
 
+#define MIN(x, y) (((x) < (y)) ? (x) : (y))
+#define MAX(x, y) (((x) > (y)) ? (x) : (y))
+
+class BoundingBox {
+public:
+    int min_x, max_x, min_y, max_y;
+
+    BoundingBox(): min_x(-1), max_x(-1), min_y(-1), max_y(-1) {}
+
+    bool empty() const {
+        return min_x == -1;
+    }
+
+    void add(int x1, int y1, int w, int h) {
+        int x2 = x1 + w - 1, y2 = y1 + h - 1;
+        min_x = (min_x < 0) ? x1 : MIN(min_x, x1);
+        min_y = (min_y < 0) ? y1 : MIN(min_y, y1);
+        max_x = (max_x < 0) ? x2 : MAX(max_x, x2);
+        max_y = (max_y < 0) ? y2 : MAX(max_y, y2);
+    }
+
+    bool intersets(const BoundingBox& other) const {
+        return !(other.min_x > max_x ||
+                 other.max_x < min_x ||
+                 other.min_y > max_y ||
+                 other.max_y < min_y);
+    }
+
+    bool tryMerge(BoundingBox& other) {
+        if (!intersets(other)) return false;
+
+        min_x = MIN(min_x, other.min_x);
+        min_y = MIN(min_y, other.min_y);
+        max_x = MAX(max_x, other.max_x);
+        max_y = MAX(max_y, other.max_y);
+        other.min_x = other.max_x = other.min_y = other.max_y = -1;
```

Comment:
this is not obvious for `tryMerge` name :confused: - need to look at usage to find out why it is here

Reply:
right, fixing